Optimize Nextcloud installation - Phase I

MichaIng opened the pull request:
https://github.com/Fourdee/DietPi/issues/1067

I added some other small fixes/adjustments I fell over. I successfully tested this configuration now for some months, but careful testing should be done, especially for Jessie, before merging.

I will go on with the other topics mentioned in the issue above. Would appreciate if someone could help with the nginx configuration: https://docs.nextcloud.com/server/12/admin_manual/installation/nginx.html

Commit:
Optimize Nextcloud installation - Phase I

https://github.com/Fourdee/DietPi/issues/1067

1. Fix wording
2. Adjust MySQL/MariaDB installation:
- mysql-server's dependency chain includes the related client for jessie and stretch, thus the client meta package doesn't need to be selected: https://packages.debian.org/de/jessie/mysql-client
3. Fix PHP MySQL module for Stretch:
- For distro Stretch and above, php(7.0)-mysqlnd does not exist, thus php(7.0)-mysql need to be installed.
4. Make sure all necessary Apache modules are installed in case.
5. Make sure all necessary PHP modules are installed.
6. Create and enable Nextcloud specific Apache configuration file with included directory specific PHP configuration.

Commit:
Disable and remove newly created nextcloud.conf on Nextcloud removal.

Fourdee commented on dietpi/dietpi-software:
```diff
@@ -8220,7 +8228,71 @@ _EOF_
+php_admin_value opcache.enable_cli 1
+php_admin_value opcache.interned_strings_buffer 8
+php_admin_value opcache.max_accelerated_files 10000
+php_admin_value opcache.memory_consumption 128
```
We set OPcache already with: https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-software#L7986-L8007 based on RAM on system.

So this entry should be removed.

Fourdee replied:
Can check Opcache size with:
http://IP_ADDRESS/opcache.php

MichaIng replied:
The thing is that Nextcloud at the moment produces a warning on admin panel, if you choose opcache settings different than this. For that reason I added the settings inside directory directive of apache conf instead of usual php.ini or opcache.ini. That way the cache size is just set locally for the nextcloud folder and for all other web sites/services the global setting from dietpi-software will stay valid.

Fourdee replied:
@MichaIng 

> if you choose opcache settings different than this. For that reason I added the settings inside directory directive of apache conf instead of usual php.ini or opcache.ini.

Ok 👍 

But we may have issues with 256MB RAM users and a 128MB cache. Most of our users install multiple software titles, but with the swapfile increase recently, should be fine.

MichaIng replied:
I just switched my opcache.memory_consumption to 512MB to see the effect on actual memory usage by htop (and will leave it like this for a while). Actually there didn't change much. As there is a limited amount php files of nextcloud, the actual opcache usage can't grow way larger than ~30MB, even if you install all apps, available at nextcloud appstore. PHP files simply do not grow that large usually (compared to images etc.).* For that reason actual memory overload is highly unlikely.

*This is also the reason why I want to start a discussion on nextcloud side to reduce or even remove that setting from their checks. It simply highly depends on additional usage of the webserver. But as for now, I think we should keep the 128MB, as a warning on nextcloud admin panel will result in reports/questions either here or on help.nextcloud.com.

Fourdee reviewed (commented):
Looks good, 1 issue with opcache, we already set it, if we can get that line removed i'll merge and test.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8220,7 +8228,71 @@ _EOF_
+			#Create Nextcloud specific webserver + PHP config.
+			#Apache: https://docs.nextcloud.com/server/12/admin_manual/installation/source_installation.html#apache-web-server-configuration
+			if (( ${aSOFTWARE_INSTALL_STATE[83]} >= 1 )); then
+				cat << _EOF_ > /etc/apache2/sites-available/nextcloud.conf
```
Should we add `nextcloud.conf` to `dietpi/conf/` and just copy it here? This would keep the already large grown `dietpi-software` a bid smaller. I already have issues when trying to watch it with mobile 😆.

Fourdee replied:
@MichaIng 

> Should we add nextcloud.conf to dietpi/conf/ and just copy it here?

The original reason for keeping everything in ```dietpi-software```, reduces space of RAMdisk usage ```/DietPi``` as each file (regardless of size) uses at least 4K, where as its probably smaller integrated into ```dietpi-software```.

> I already have issues when trying to watch it with mobile

Yep, shes a big one alright lol.
By all means, if you want to ```cp``` from ```/DietPi/dietpi/conf/nextcloud.conf``` instead, go for it 👍

MichaIng replied:
Yeah I see, for optimization reasons, also in case of RAM usage, we should keep all config files < 4K inside of dietpi-software.

The only question then is: Why do we keep the the conf/ actually inside ramdisk? Would be a separate project to switch all of them together with all necessary path changes, but on the long term we could choose another folder to put the config files inside, that is outside of ramdisk. This would then reduce RAM usage by the amount, dietpi-software becomes slimmer 🙂.

Commit:
Install php(5)-intl module before trying to enable it

jbz-dev commented:
Hi @MichaIng 

I've just reinstalled Nextcloud and here is the working nginx conf for it (the one that is on their website)

```
    location = /.well-known/carddav {
      return 301 $scheme://$host/nextcloud/remote.php/dav;
    }
    location = /.well-known/caldav {
      return 301 $scheme://$host/nextcloud/remote.php/dav;
    }

    location /.well-known/acme-challenge { }

    location ^~ /nextcloud {

        # set max upload size
        client_max_body_size 512M;
        fastcgi_buffers 64 4K;

        # Enable gzip but do not remove ETag headers
        gzip on;
        gzip_vary on;
        gzip_comp_level 4;
        gzip_min_length 256;
        gzip_proxied expired no-cache no-store private no_last_modified no_etag auth;
        gzip_types application/atom+xml application/javascript application/json application/ld+json application/manifest+json application/rss+xml application/vnd.geo+json application/vnd.ms-fontobject application/x-font-ttf application/x-web-app-manifest+json application/xhtml+xml application/xml font/opentype image/bmp image/svg+xml image/x-icon text/cache-manifest text/css text/plain text/vcard text/vnd.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy;

        # Uncomment if your server is build with the ngx_pagespeed module
        # This module is currently not supported.
        #pagespeed off;

        location /nextcloud {
            rewrite ^ /nextcloud/index.php$uri;
        }

        location ~ ^/nextcloud/(?:build|tests|config|lib|3rdparty|templates|data)/ {
            deny all;
        }
        location ~ ^/nextcloud/(?:\.|autotest|occ|issue|indie|db_|console) {
            deny all;
        }

        location ~ ^/nextcloud/(?:index|remote|public|cron|core/ajax/update|status|ocs/v[12]|updater/.+|ocs-provider/.+)\.php(?:$|/) {
            fastcgi_split_path_info ^(.+\.php)(/.*)$;
            include fastcgi_params;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            fastcgi_param PATH_INFO $fastcgi_path_info;
            fastcgi_param HTTPS on;
            #Avoid sending the security headers twice
            fastcgi_param modHeadersAvailable true;
            fastcgi_param front_controller_active true;
            fastcgi_pass php-handler;
            fastcgi_intercept_errors on;
            fastcgi_request_buffering off;
        }

        location ~ ^/nextcloud/(?:updater|ocs-provider)(?:$|/) {
            try_files $uri/ =404;
            index index.php;
        }

        # Adding the cache control header for js and css files
        # Make sure it is BELOW the PHP block
        location ~ \.(?:css|js|woff|svg|gif)$ {
            try_files $uri /nextcloud/index.php$uri$is_args$args;
            add_header Strict-Transport-Security "max-age=15768000; includeSubDomains; preload;";
            add_header Cache-Control "public, max-age=15778463";
	    add_header X-Content-Type-Options nosniff;
            add_header X-XSS-Protection "1; mode=block";
            add_header X-Robots-Tag none;
            add_header X-Download-Options noopen;
            add_header X-Permitted-Cross-Domain-Policies none;
            # Optional: Don't log access to assets
            access_log off;
        }

        location ~ \.(?:png|html|ttf|ico|jpg|jpeg)$ {
            try_files $uri /nextcloud/index.php$uri$is_args$args;
            # Optional: Don't log access to other assets
            access_log off;
        }

```

The configuration is located in `/etc/nginx/sites-available/nextcloud.conf`, simlinked with `site-enabled/nextcloud.conf`

MichaIng commented:
Thanks for this. I will try to add the php module settings, the same way they are added into the apache config. Would be great if you could test it afterwards.

But maybe this will be delayed onto Phase II, if fourdee feels/tested the current PR state ready to be merged.

Fourdee commented:
@MichaIng 
> But maybe this will be delayed onto Phase II, if fourdee feels/tested the current PR state ready to be merged.

Yep, i'd prefer to wait until all web server's we offer have this feature in PR, before merging. Especially as Lighttpd is our default webserver. Then testing can begin.

MichaIng commented:
Okay, good to know, so I will add nginx configuration soon.

> Especially as Lighttpd is our default webserver.

According to nextcloud admin manual and doc of their dav server (sabre), lighttpd might not work (well) or missing features for nextcloud:
https://docs.nextcloud.com/server/12/admin_manual/issues/general_troubleshooting.html?highlight=lighttpd#web-server-and-php-modules
On the other hand there are some people using the combination without any major reported problem. So I would give it a shot.
As there is nothing officially provided by nextcloud: Someone there with a working configuration that covers moreless what is handles within the apache/nginx configuration files, especially disabling/excluding certain php modules and serving security headers? 😄

Commit:
Adding nginx configuration file including php module settings

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8294,6 +8288,122 @@ _EOF_
+            access_log off;
+        }
+	
+	# Setting PHP values by using this method: http://php.net/manual/en/install.fpm.configuration.php Example #2
```
@Bridouz Added the php module settings here, same as for apache. I hope it is working as intended at this place (just inside `location ^~ /nextcloud {`.
This was by the way not "closed" in your config paste, so I added `}` at the end, hope that is right also.

Would appreciate testing, also via `phpinfo()`, if the module settings are active for `/nextcloud`.

Commit:
Correct nginx PHP value assigning

Other than in apache.conf, PHP values must be set with "=" in nginx.conf.

Commit:
Remove Nextcloud nginx config on uninstall

Also suppressing error messages about missing files and commands, as only one webserver is installed. Not sure if there is a reliable way to check which webserver "was" installed, as it could be removed before or together with Nextcloud or not.

Fourdee commented:
@MichaIng 
Great work, lets merge for testing 👍

jbz-dev commented:
Installation fails on fresh Rpi3 DietPi V157 : 

Installation aborted because : 
```
E: Command line option 'i' [from -intl] is not understood in combination with the other options.
 [Failed] The apt cache may be corrupt, apt mirror offline, or you have held broken packages. DietPi-Software will now exit.
```

In fact `php7.0-intl" does not exist in apt-cache.

I tried to remove that package installation but then I had another errors with : 
```
${php_package_name}enmod ctype curl dom fileinfo gd iconv intl json mbstring posix simplexml xmlwriter xmlreader zip pdo_mysql exif opcache apcu
```
 saying that packages like `opcache` `apcu` do not exist.

EDIT: I removed all the php installation stuff locally as I already have them. Installation OK

but nginx fails to restart with error : 
```
nginx: [emerg] "location" directive is not allowed here in /etc/nginx/sites-enabled/nextcloud.conf:1
```

MichaIng commented:
Lets stay here with PR discussion/bugfix:

@Bridouz

About php-intl:
- php_package_name should be "php" on stretch and "php5" on Jessie. The related package names should exist in the debian and raspbian repo. So php7.0-intl should not be tried to install. Could you verify this again? Are you using Stretch or Jessie?
- But in fact the error looks more like the -intl part ia interpreted as command options instead as part of the package name. I will correct this.
`€: Just check again on my server: The command runs fine here on stretch, package exists, command is also used the same way above on php installation. I am little confused.`

About phpenmod:
- Did you run the whole dietpi-software? As apcu and opcache should be installed together with php outside of the nextcloud related stuff.
- Do they show up in "php -m" and did you try to install and activate them manually? The command with php(5) varible inside works fine on my raspbian stretch.
`€: opcache is no separate package on Jessie, maybe there is thus no ini file the enable/disable? APCu should work fine, if installed with php installation. Any other modules that produce error messages on Jessie?`

About nginx:
Hmm as you posted this the same way:
Does your working nextcloud.conf start the same way, with location? Or did you just copy a part of your file and the location directive is actually inside the server directive like on nextcloud manual? In that case we would need to add the config part into the existing port 80 or port 443 conf, dependent on if ssl is active and used or not. Would be annoying to realize, I hope separate config snippets with location inside are possible with nginx, as they are with apache.

jbz-dev commented:
@MichaIng 

About Nginx: I think I totally f***** up my configuration when I sent it to you few days ago. I will repost it later with also a modification on default nginx on DietPi to split server blocks for Http and https. I think that as we (I think) mostly use only one domain name and want to use nextcloud on a subdirectory (mydomain.com/nextcloud) it would great to put the `nextcloud.conf` in `sites-dietpi`.

Time to sleep and I will take care of it tomorrow.

On Php (I use Stretch) : 

- yes, i think it's mainly a syntax issue for php-intl.
- `php -m` shows them
